llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -369,14 +409,160 @@ int InterBlockScheduling::getNumEntryNops(const BlockState &BS) const {
+        // associated boundary nodes, so the mapping is not available.
+        if (Succ->isBoundaryNode()) {
+          // We can consider the maximum DistFromLoopEntry as depth of the
+          // ExitSU.
+          Distance = DistFromLoopEntry - DistancesFromLoopEntry[PreBoundaryMI];
+          DEBUG_LOOPAWARE(dbgs() << "Sched boundary dependency found:\n"
+                                 << " Loop instruction: " << *PreBoundaryMI
+                                 << "\n Latency: " << Latency
+                                 << "\n Distance: " << Distance << "\n");
+
+        } else {
+          const MachineInstr *PostBoundaryMI = Succ->getInstr();
+          Distance = DistancesFromLoopEntry[PostBoundaryMI] -
+                     DistancesFromLoopEntry[PreBoundaryMI];
+          DEBUG_LOOPAWARE(dbgs() << "Data dependency found:\n"
+                                 << " Loop instruction: " << *PreBoundaryMI
+                                 << " Epilogue instruction: " << *PostBoundaryMI
+                                 << "\n Latency: " << Latency
+                                 << "\n Distance: " << Distance << "\n");
+        }
+
+        EntryNops = std::max(EntryNops, Latency - Distance);
+      }
+    }
+    if (++Height > HR->getConflictHorizon()) {
+      break;
+    }
+  }
+
+  return EntryNops;
+}
+
+int InterBlockScheduling::getCyclesToAvoidResourceConflicts(
+    int ExistingLatency, const BlockState &EpilogueBS,
+    const BlockState &LoopBS) const {
+
+  const MachineBasicBlock &EpilogueMBB = *EpilogueBS.TheBlock;
+  MachineBasicBlock *LoopMBB = LoopBS.TheBlock;
+  int Depth = HR->getMaxLookAhead();
+  ResourceScoreboard<FuncUnitWrapper> Bottom;
+  Bottom.reset(Depth);
+
+  DEBUG_LOOPAWARE(dbgs() << "* Loop/Epilogue-carried resource conflicts:"
+                         << " Original Loop " << *LoopMBB << " Original Epilog "
+                         << EpilogueMBB << "\n");
+
+  emitBundlesInScoreboard(LoopBS.getBottom().Bundles, Bottom, HR.get());
+
+  // We know how many latency cycles we need to respect, and we can advance
+  // the scoreboard to the first possible cycle that can accommodate another
+  // instruction and start the resource verification from this point, tracking
+  // the number of NOPS.
+  int NopCounter = 0;
+  for (NopCounter = 0; NopCounter < ExistingLatency; ++NopCounter)
+    Bottom.advance();
+
+  DEBUG_LOOPAWARE(dbgs() << "Loop scoreboard\n"; Bottom.dump());
+
+  ResourceScoreboard<FuncUnitWrapper> Top;
+  Top.reset(Depth);
+  int Cycle = -Depth;
+
+  auto Bundles = EpilogueBS.getBottom().Bundles;
+
+  emitBundlesInScoreboardDelta(EpilogueBS.getBottom().Bundles, Top, Cycle,
+                               HR.get());
+
+  DEBUG_LOOPAWARE(dbgs() << "Epilogue scoreboard\n"; Top.dump());
+
+  // Use scoreboard comparison to calculate the number of nops
+  while (Bottom.conflict(Top, Depth)) {
+    Bottom.advance();
+    NopCounter++;
+  }
```

Comment:
Nit: That looks a lot like `AIEPostRASchedStrategy::handleRegionConflicts`, maybe there's code we can share.

Reply:
Hi @gbossu, I think it could be a bit complicated, because `handleRegionConflicts` uses two hazard recognizers, while we use just one, plus scoreboard comparison. I am afraid that a refactor could create a more confusing code.

Reply:
The implementation is different, but ultimately the goal is the same, we want to insert NOPs until there is no resource hazard. We can do that in a follow-up PR.